.circleci/config.yml

```diff
@@ -1,240 +1,396 @@
+          sha256sum | \
+          awk '{print "overlay_dependencies "$1}' >> $OVERLAY_WS/checksum.txt
+  store_overlay_checksum: &store_overlay_checksum
     store_artifacts:
-        path: checksum.txt
-  restore_build_cache: &restore_build_cache
+      path: checksum.txt
+  restore_overlay_cache: &restore_overlay_cache
     restore_cache:
-      name: Restore Build Cache
-      key: build-cache-v1-{{ arch }}-{{ .Branch }}-{{ checksum "/opt/nav2_ws/checksum.txt" }}
-  source_build: &source_build
+      name: Restore Overlay Cache
+      key: "overlay-v1\
+        -{{ arch }}\
+        -{{ .Branch }}\
+        -{{ .Environment.CIRCLE_PR_NUMBER }}\
+        -{{ checksum \"/opt/overlay_ws/checksum.txt\" }}"
+  build_overlay: &build_overlay
     run:
-      name: Source Build
+      name: Build Overlay
       command: |
-        if [ -d install ]
+        if [ -d install ] && [ ! -f build_failed ]
         then
-            echo "Skipping Source Build"
+            echo "Skipping Overlay Build"
         else
-            . $ROS_WS/install/setup.sh
+            rm -rf build install log
+            . $UNDERLAY_WS/install/setup.sh
             colcon build \
               --symlink-install \
               --cmake-args \
                 -DCMAKE_BUILD_TYPE=$CMAKE_BUILD_TYPE \
+                -DCMAKE_C_COMPILER_LAUNCHER=ccache \
+                -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                 -DCOVERAGE_ENABLED=$COVERAGE_ENABLED
+            rm -f build_failed
         fi
-  save_build_cache: &save_build_cache
+  store_ccache_logs: &store_ccache_logs
+    store_artifacts:
+        path: /tmp/ccache.log
+  save_ccache: &save_ccache
     save_cache:
-      name: Save Build Cache
-      key: build-cache-v1-{{ arch }}-{{ .Branch }}-{{ checksum "/opt/nav2_ws/checksum.txt" }}
+      name: Save CCache
+      key: "ccache-v1\
+        -{{ arch }}\
+        -{{ .Branch }}\
+        -{{ .Environment.CIRCLE_PR_NUMBER }}\
+        -{{ checksum \"/opt/underlay_ws/checksum.txt\" }}\
+        -{{ epoch }}"
       paths:
-        - /opt/nav2_ws
+        - /root/.ccache
+      when: always
+  save_overlay_cache: &save_overlay_cache
+    save_cache:
+      name: Save Overlay Cache
+      key: "overlay-v1\
+        -{{ arch }}\
+        -{{ .Branch }}\
+        -{{ .Environment.CIRCLE_PR_NUMBER }}\
+        -{{ checksum \"/opt/overlay_ws/checksum.txt\" }}\
+        -{{ epoch }}"
+      paths:
+        - /opt/overlay_ws
   copy_build_logs: &copy_build_logs
     run:
-        name: Copy Build Logs
-        when: always
-        command: cp -rH log/latest_build log/build
+      name: Copy Build Logs
+      when: always
+      command: cp -rH log/latest_build log/build
   store_build_logs: &store_build_logs
     store_artifacts:
         path: log/build
-  test_build: &test_build
+  # test_overlay: &test_overlay
+  #   run:
+  #     name: Test Overlay
+  #     command: |
+  #       . install/setup.sh
+  #       src/navigation2/tools/run_test_suite.bash
```

Comment:
Perhaps we could push any system integration tests into proper ctests  that include timing info in the xml test results, so that they can be load balanced accordingly. We could also split them out into their own job, but I'd like to make sure they are still reporting their own timing.

Reply:
> Perhaps we could push any system integration tests into proper ctests

They are are ctests already. Do you mean launch them from `colcon test`?

Reply:
> Do you mean launch them from colcon test

Yea, as then they could be load balanced by the circleci timing split. However splitting currently is at the granularity of package, so if one package take the lion share of test time, say more than the sum of the other tests; one unlucky worker will still end up testing that package by its lonesome.

Reply:
Ok. I don't want to go back to colcon running the system tests directly without figuring out some changes. I was unable to get the full console output when colcon ran the test, and without that, it was impossible to tell why the test failed.

Reply:
> I was unable to get the full console output when colcon ran the test

Could you be more specific? Do you have an example I could debug, or links to the two CI test runs I could compare?

Reply:
I just ran the system test locally using colcon. Here is the output:
```
> colcon test --merge-install --packages-select nav2_system_tests
Starting >>> nav2_system_tests
[Processing: nav2_system_tests]                   
[Processing: nav2_system_tests]                           
[Processing: nav2_system_tests]                             
[Processing: nav2_system_tests]                             
--- stderr: nav2_system_tests                               
Errors while running CTest
---
Finished <<< nav2_system_tests [2min 30s]	[ with test failures ]

Summary: 1 package finished [2min 30s]
  1 package had stderr output: nav2_system_tests
  1 package had test failures: nav2_system_tests

> 

> colcon test-result --verbose --test-result-base build/nav2_system_tests/
build/nav2_system_tests/Testing/20190814-2300/Test.xml: 11 tests, 0 errors, 1 failure, 0 skipped
build/nav2_system_tests/test_results/nav2_system_tests/test_localization.xml: 1 test, 0 errors, 1 failure, 0 skipped
- nav2_system_tests test_localization.missing_result
  <<< failure message
    The test did not generate a result file.
  >>>

Summary: 77 tests, 0 errors, 2 failures, 0 skipped

```

That's all I get - "test did not generate a result file"

Compare with any current build where I get all the console output - showing all the messages each node prints out.

Reply:
That summary is being generated with colcon by exampling what ctest wrote to [test_results/nav2_system_tests/test_localization.xml](https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/test_results/nav2_system_tests/test_localization.xml). 

I'm not sure why the ctest isn't reporting the stdout/stderr. My guess is that it wasn't configured to, but perhaps Dirk would know. Still, I don't think it would be useful to duplicate the stdout dump into the xml when its already captured from the log artfacts. I think that message field in the xml should be used for heigherlevel feedback reporting.

But looking at the colcon logs for the test, all the console output - showing all the messages each node prints out - seems to be there: logs/test/nav2_system_tests/... Please make sure that you can navigatiate the artifact tab for test jobs where this is all stored for you.

https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/stdout.log
https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/stdout_stderr.log
https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/streams.log
https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/command.log

![image](https://user-images.githubusercontent.com/2293573/63077065-edef8480-beeb-11e9-9795-4cbe5f861e3e.png)